pkg/action/cl001/ac008/executor.go

```diff
@@ -2,8 +2,279 @@ package ac008
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if len(kiamServerPods.Items) != expectedKiamServerPodCount {
+			e.logger.LogCtx(ctx, "level", "error", "message", fmt.Sprintf("Wrong kiam-server pod count, expected %d but got %d", expectedKiamServerPodCount, len(kiamServerPods.Items)))
+			return microerror.Mask(executionFailedError)
+		}
+
+		for _, kiamServerPod := range kiamServerPods.Items {
+			if kiamServerPod.Status.Phase != apiv1.PodRunning {
+				e.logger.LogCtx(ctx, "level", "error", "message", fmt.Sprintf("Pod %s in namespace %s is not running.", kiamServerPod.Name, kiamServerPod.Namespace))
+				return microerror.Mask(executionFailedError)
+			}
+		}
+	}
+
+	// kiam agent
+	{
+		kiamAgentPods, err := tcClient.CoreV1().Pods(kubeSystemNamespace).List(ctx, metav1.ListOptions{LabelSelector: kiamAgentLabelSelector})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if len(kiamAgentPods.Items) != expectedKiamAgentPodCount {
+			e.logger.LogCtx(ctx, "level", "error", "message", fmt.Sprintf("Wrong kiam-agent pod count, expected %d but got %d", expectedKiamAgentPodCount, len(kiamAgentPods.Items)))
+			return microerror.Mask(executionFailedError)
+		}
+
+		for _, kiamAgentPod := range kiamAgentPods.Items {
+			if kiamAgentPod.Status.Phase != apiv1.PodRunning {
+				e.logger.LogCtx(ctx, "level", "error", "message", fmt.Sprintf("Pod %s in namespace %s is not running.", kiamAgentPod.Name, kiamAgentPod.Namespace))
+				return microerror.Mask(executionFailedError)
+			}
+		}
+	}
+
+	return nil
+}
+
+// testAWSApiCalls will spawn a job in k8s tenant cluster to test calling AWS API to ensure kiam works as expected
+func (e *Executor) testAWSApiCalls(ctx context.Context, tcClient kubernetes.Interface, awsRegion string, dockerRegistry string) error {
+	networkPolicy := jobNetworkPolicy()
+	_, err := tcClient.NetworkingV1().NetworkPolicies(kubeSystemNamespace).Create(ctx, networkPolicy, metav1.CreateOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	// delete the network policies when function ends
+	defer func() {
+		_ = tcClient.NetworkingV1().NetworkPolicies(kubeSystemNamespace).Delete(ctx, networkPolicy.Name, metav1.DeleteOptions{})
+	}()
+
+	job := awsApiCallJob(dockerRegistry, awsRegion, e.tenantCluster)
+	_, err = tcClient.BatchV1().Jobs(kubeSystemNamespace).Create(ctx, job, metav1.CreateOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	// delete the job when function ends
+	defer func() {
+		_ = tcClient.BatchV1().Jobs(kubeSystemNamespace).Delete(ctx, job.Name, metav1.DeleteOptions{})
+	}()
+
+	// check if job is completed
+	o := func() error {
+		job, err := tcClient.BatchV1().Jobs(kubeSystemNamespace).Get(ctx, job.Name, metav1.GetOptions{})
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if !isJobCompleted(job) {
+			return microerror.Mask(jobNotCompleted)
+		}
+		return nil
+	}
+	b := backoff.NewMaxRetries(checkJobRetryLimit, checkJobMaxWait)
```

Comment:
Retrying in the action feels kind of wrong. I am wondering about the whole test design. I am sure this can work in many different ways, though the idea of the action concept in `awscnfm` was to have many of them and let each action do one thing well. Now for this "kiam test" there seem to have to happen a lot of different things. Can we split this up? The problem I see is that when you execute this new action here now there are potentially 10 different reasons why the action might fail. The idea for actions was that when an action fails, you know what to look at, because an action is supposed to do only one thing. When this action here now fails, nobody knows why.

Reply:
I need to think about how would that work because splitting into 2 actions just because of 10-20s wait feels weird since it would mean tons of code duplication

Reply:
It is not about waiting. The action to verify if the cluster upgrade is successful can take up to 2 hours to complete. Though the action itself does one thing and then it fails. When it fails the plan execution puts a backoff onto it. The action itself does not retry. The action itself acts and comes to a conclusion real quick. When we have now an action that does 10 different things and in between also retries something the whole event chain of this new action here will be started over again once anything in the middle breaks. At this point we are not so much concerned with duplicated code. We generate most of the boilerplate already anyway.